Snowflake.Data/Core/SFBaseResultSet.cs

```diff
@@ -43,28 +117,88 @@ internal string GetString(int columnIndex)
             switch (type)
             {
                 case SFDataType.DATE:
-                    var val = GetValue(columnIndex);
+                    var val = GetValue(columnIndex, out double GetValueTime);
                     if (val == DBNull.Value)
                         return null;
-                    return SFDataConverter.toDateString((DateTime)val, 
+                    var result = Measure(
+                        @"SFDataConverter.toDateString", 
+                        out double toDateStringTime,
+                        () =>
+                    {
+                        return SFDataConverter.toDateString((DateTime)val,
                         sfResultSetMetaData.dateOutputFormat);
+                    });
+
+                    // Total time for GetString
+                    addTimesToTimingMaps("GetString", GetValueTime + toDateStringTime);
+
+                    return result;
                 //TODO: Implement SqlFormat for timestamp type, aka parsing format specified by user and format the value
                 default:
-                    return getObjectInternal(columnIndex); 
+                    return Measure(@"getObjectInternal", out double getObjectInternalTime, () =>
+                    {
+                        return getObjectInternal(columnIndex);
+                    });
             }
         }
 
-        internal object GetValue(int columnIndex)
+        internal object GetValue(int columnIndex, out double time)
         {
-            string val = getObjectInternal(columnIndex);
+            string val = Measure(@"getObjectInternal", out double getObjectInternalTime, () =>

```

Comment:
Do you know the overhead of this Measure function?

Reply:
It is quite big. The run time increased from 10 to 15sec in the sample query "select * from SNOWFLAKE_SAMPLE_DATA.TPCH_SF100.LINEITEM limit 350000".

Reply:
This seems like a very bad idea!

SFBaseResultSet.GetValue() is a very performance-sensitive function. It can be called millions of times. You should absolutely not add any extra overhead to this code path.

I have used the following code to test:

```
            using (IDbCommand command = conn.CreateCommand())
            {
                command.CommandText = "select * from SNOWFLAKE_SAMPLE_DATA.TPCH_SF100.LINEITEM limit 1500000";
                var dr = command.ExecuteReader();
                sw = System.Diagnostics.Stopwatch.StartNew();
                while (dr.Read())
                {
                    rowcount++;
                    for (int i = 0; i < dr.FieldCount; i++)
                    {
                        dr.GetValue(i);
                    }
                }
                sw.Stop();
                Console.WriteLine("rowcount = {0} Time = {1}", rowcount, sw.Elapsed);
            }

```
I have tested this from a console program running NetFramework with all logging turned OFF

After this change: 42.1 s
Before this change: 18.0 s

This change makes the code more than twice as slow in an important usecase even without turning on logging at all.

If we do the same test with logLevel=DEBUG we get 61 s instead.

Please do NOT merge this PR.

Reply:
Hi Stefan, 
Don't worry, we are not going to ever merge this!
We will be using this to collect more detailed metric on the different types conversions and only timing the retrieval execution as you are doing here does not help with this. Yes, when data conversion is improved it shows up in the overall time, but it doesn't highlight if a specific conversion (JSON to C#) need the most improvement or if the raw data transfer is a bottle neck.
I know you worked on improving the conversion to numeric types and this is pretty awesome, but there is most likely other things that can be improved.

Reply:
Good to hear that you are not planning to merge this!

I am not really sure what you are trying to accomplish here, but this kind of micro-benchmarking can be very misleading. 

You try to measure each individual function call and then sum all the measurements. This does not work properly for these very fast functions.

On my computer, Stopwatch has a resolution of 100 ns - each tick is 100ns

dr.GetValue(0) for an int64 column (with my improvements) takes about 80 ns (measured by calling it 10000000 times in a tight loop).

It is impossible to measure something that only takes 80ns with a clock that has 100ns resolution. You are only measuring your own overhead.

If you want to measure the performance of GetValue() for different kinds of data, I recommend that you call GetValue() in a tight loop without calling dr.Next().

That way you are not loading any data from snowflake while measuring performance.

You might also want to have a look at this: https://github.com/dotnet/BenchmarkDotNet

Reply:
Thanks Stefan! BenchmarkDotNet seems very good for benchmarking specific methods like the conversion method. I didn't see how it could be applied into gathering metrics from a customer application, but it is definitely useful.